backend/zkppSalt.js

```diff
@@ -0,0 +1,254 @@
+
+// TODO HARDCODED VALUES
+// These values will eventually come from server the configuration
+const recordPepper = 'pepper'
+const recordMasterKey = 'masterKey'
+const challengeSecret = 'secret'
+const registrationSecret = 'secret'
+const maxAge = 30
+
+const getZkppSaltRecord = async (contract: string) => {
+  const recordId = blake32Hash(hashStringArray('RID', contract, recordPepper))
+  const record = await sbp('chelonia/db/get', recordId)
+
+  if (record) {
+    const encryptionKey = hashStringArray('REK', contract, recordMasterKey).slice(0, nacl.secretbox.keyLength)
+
+    const recordBuf = Buffer.from(base64urlToBase64(record), 'base64')
+    const nonce = recordBuf.slice(0, nacl.secretbox.nonceLength)
+    const recordCiphertext = recordBuf.slice(nacl.secretbox.nonceLength)
+    const recordPlaintext = nacl.secretbox.open(recordCiphertext, nonce, encryptionKey)
+
+    if (!recordPlaintext) {
+      return null
+    }
+
+    const recordString = Buffer.from(recordPlaintext).toString('utf-8')
+
+    try {
+      const recordObj = JSON.parse(recordString)
+
+      if (!Array.isArray(recordObj) || recordObj.length !== 3 || !recordObj.reduce((acc, cv) => acc && typeof cv === 'string', true)) {
+        console.error('Error validating encrypted JSON object ' + recordId)
+        return null
+      }
+
+      const [hashedPassword, authSalt, contractSalt] = recordObj
+
+      return {
+        hashedPassword,
+        authSalt,
+        contractSalt
+      }
+    } catch {
+      console.error('Error parsing encrypted JSON object ' + recordId)
+      // empty
+    }
+  }
+
+  return null
+}
+
+const setZkppSaltRecord = async (contract: string, hashedPassword: string, authSalt: string, contractSalt: string) => {
+  const recordId = blake32Hash(hashStringArray('RID', contract, recordPepper))
+  const encryptionKey = hashStringArray('REK', contract, recordMasterKey).slice(0, nacl.secretbox.keyLength)
+  const nonce = nacl.randomBytes(nacl.secretbox.nonceLength)
+  const recordPlaintext = JSON.stringify([hashedPassword, authSalt, contractSalt])
+  const recordCiphertext = nacl.secretbox(Buffer.from(recordPlaintext), nonce, encryptionKey)
+  const recordBuf = Buffer.concat([nonce, recordCiphertext])
+  const record = base64ToBase64url(recordBuf.toString('base64'))
+  await sbp('chelonia/db/set', recordId, record)
+}
+
+export const getChallenge = async (contract: string, b: string): Promise<false | {authSalt: string; s: string; sig: string;}> => {
+  const record = await getZkppSaltRecord(contract)
+  if (!record) {
+    console.debug('getChallenge: Error obtaining ZKPP salt record for contract ID ' + contract)
+    return false
+  }
+  const { authSalt } = record
+  const s = randomNonce()
+  const now = (Date.now() / 1000 | 0).toString(16)
+  const sig = [now, base64ToBase64url(Buffer.from(hashStringArray(contract, b, s, now, challengeSecret)).toString('base64'))].join(',')
```

Comment:
Same question re `.toString('base64url')` here

Reply:
Not sure I can find the original question. But if the question is why `base64ToBase64url` instead of `.toString('base64url')`, that's because `Buffer` is not standard JS and the shim that we're using doesn't support it.